core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -1085,39 +1534,49 @@ public void purge(Executor threadPool, final PurgeListener task) {
             }
          }
       }
+      return entriesToPurge;
+   }
+
+   @Override
+   public CompletionStage<Long> size(IntSet segments) {
+      return Flowable.fromPublisher(publishKeys(segments, null)).count().toCompletionStage();
+   }
 
+   @Override
+   public CompletionStage<Long> approximateSize(IntSet segments) {
+      if (!segmented && segments.size() < ctx.getCache().getCacheConfiguration().clustering().hash().numSegments()) {
+         return Flowable.fromPublisher(publishKeys(segments, null)).count().toCompletionStage();
+      }
+      return blockingManager.supplyBlocking(() -> blockingApproximateSize(segments), "sfs-approximateSize");
+   }
+
+   private long blockingApproximateSize(IntSet segments) {
       long stamp = resizeLock.readLock();
+      long size = 0;
       try {
-         for (Iterator<KeyValuePair<Object, FileEntry>> it = entriesToPurge.iterator(); it.hasNext(); ) {
-            KeyValuePair<Object, FileEntry> next = it.next();
-            FileEntry fe = next.getValue();
-            if (fe.isExpired(now)) {
-               it.remove();
-               try {
-                  free(fe);
-               } catch (Exception e) {
-                  throw new PersistenceException(e);
-               }
-               if (task != null) task.entryPurged(next.getKey());
+         for (int segment = 0; segment < actualNumSegments; segment++) {
+            Map<K, FileEntry> segmentEntries = getSegmentEntries(segment);
+            if (segmentEntries != null && segments.contains(segment)) {
```

Comment:
```suggestion
         for (PrimitiveIterator.OfInt iter = segments.iterator(); iter.hasNext(); ) {
            Map<K, FileEntry> segmentEntries = getSegmentEntries(iter.next());
            if (segmentEntries != null) {
```

I don't see why this wouldn't work instead or are you trying to avoid an allocation of the primitive iterator?

Reply:
Yeah, it should work